### PR TITLE
Crafting Fixes

### PR DESCRIPTION
**Pull request by JamesGreen31**
Resolved issue 229 relating to crafting issues and improved the overall health of the crafting with recipie book system.

**Issues resolved:**

1. Resolved an issue where crafting with the crafting book would sometimes result in getting stuck endlessly switching between items.
2. Resolved an issue where crafting with the crafting book would break at a crafting table (would get super spammy with opening and closing the table). This was caused by items getting stuck in the 2x2 player crafting grid.

**Update Notes:**

The fixes in this update only apply to crafting with the recipe book. Any issues with crafting that do not deal with that are still unresolved. 

The ender eye issue is not necessarily a crafting issue since the task works fine by itself. Rather, it is an issue with the tasks interrupting each other when running BeatMinecraft2. I will investigate this in the future, but I wanted to get these crafting fixes out now.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -117,3 +117,4 @@ run/
 
 # Avoid ignoring Gradle wrapper jar file (.jar files are usually ignored)
 !gradle-wrapper.jar
+
```
Comment (JamesGreen31) [inline .gitignore:120]: no real changes, just some spacing from when I tried to exclude a few documents but gave up.

File: a [hunk 2]
```diff
@@ -0,0 +1,487 @@
+[33mcommit e4195ceda64f9844b686df14ab04d22a96ff79e5[m[33m ([m[1;36mHEAD -> [m[1;32mCraftingbookFix[m[33m, [m[1;31morigin/CraftingbookFix[m[33m)[m
```
Comment (JamesGreen31) [inline a:1]: this is garbage. We should add this to the gitignore file later.

File: src/main/java/adris/altoclef/control/SlotHandler.java [hunk 3]
```diff
@@ -48,7 +48,6 @@ public void registerSlotAction() {
         _slotActionTimer.reset();
     }
 
-
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/control/SlotHandler.java:51]: no changes

File: src/main/java/adris/altoclef/tasks/ResourceTask.java [hunk 4]
```diff
@@ -33,7 +38,7 @@
  *
  * If the target item is on the ground or in a chest, will grab from those sources first.
  */
-public abstract class ResourceTask extends Task implements ITaskCanForce {
+public abstract class ResourceTask extends Task implements ITaskCanForce, ITaskLimitsSlots {
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/ResourceTask.java:41]: implement new interface

File: src/main/java/adris/altoclef/tasks/ResourceTask.java [hunk 5]
```diff
@@ -176,6 +179,16 @@ protected Task onTick(AltoClef mod) {
             }
         }
 
+        return onResourceTick_clearingPhase(mod);
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/ResourceTask.java:182]: define methods from new interface.

File: src/main/java/adris/altoclef/tasks/ResourceTask.java [hunk 6]
```diff
@@ -247,6 +260,23 @@ public ResourceTask forceDimension(Dimension dimension) {
 
     protected abstract String toDebugStringName();
 
+    @Override
+    public boolean shouldEmptyCursorSlot(AltoClef mod, Task candidate) {
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/ResourceTask.java:264]: this is here if we need it in the future

File: src/main/java/adris/altoclef/tasks/ResourceTask.java [hunk 7]
```diff
@@ -247,6 +260,23 @@ public ResourceTask forceDimension(Dimension dimension) {
...
+        return false;
+    }
+
+    @Override
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/ResourceTask.java:268]: This method ensures that the 2x2 crafting grid is always clear so items cannot get stuck. The only time this doesn't happen is if we are crafting in our inventory.

File: src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
+package adris.altoclef.tasks.slot;
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java:1]: Task that clears out the 2x2 grid

File: src/main/java/adris/altoclef/tasks/slot/EnsureFreeCursorSlotTask.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
+package adris.altoclef.tasks.slot;
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/slot/EnsureFreeCursorSlotTask.java:1]: task which ensures the cursor slot is empty

File: src/main/java/adris/altoclef/tasks/slot/MoveInaccessibleItemToInventoryTask.java [hunk 10]
```diff
@@ -40,6 +40,12 @@ protected Task onTick(AltoClef mod) {
             if (_target.matches(StorageHelper.getItemStackInCursorSlot().getItem())) {
                 slotToMove = Optional.of(CursorSlot.SLOT);
             }
+    // issue is a full cursor slot when trying to clear out bad items.
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/slot/MoveInaccessibleItemToInventoryTask.java:43]: This resolves the bug where the bot endlessly switches between items in an attempt to clear the 2x2 grid. We handled clearing the grid elsewhere, but this needs to happen also.

File: src/main/java/adris/altoclef/tasks/slot/ReceiveOutputSlotTask.java [hunk 11]
```diff
@@ -34,25 +34,9 @@ protected void onStart(AltoClef mod) {
...
-                Optional<Slot> garbage = StorageHelper.getGarbageSlot(mod);
-                if (garbage.isPresent()) {
-                    setDebugState("Picking up garbage");
-                    return new ClickSlotTask(garbage.get());
-                } else {
-                    setDebugState("STUCK! Everything's protected.");
-                    return null;
-                }
-            }
+        boolean cursorSlotFree = cursor.isEmpty();
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasks/slot/ReceiveOutputSlotTask.java:37]: We generalized clearing the cursor slot, which is what the removed code does. Replaced with EnsureFreeCursorSlotTask

File: src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package adris.altoclef.tasksystem;
```
Comment (JamesGreen31) [inline src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java:1]: New Interface which when implemented, allows for the clearing of certain slots. Used in ResourceTask to clear the 2x2 grid when not crafting with it.

**Comment by JamesGreen31** [comment]
DO NOT MERGE YET

File: src/main/java/adris/altoclef/tasks/ResourceTask.java (hunk 4 above)
Comment (adrisj7) [reply src/main/java/adris/altoclef/tasks/ResourceTask.java:38]: Why do we need this interface?

File: src/main/java/adris/altoclef/tasks/CraftInInventoryTask.java [hunk 13]
```diff
@@ -76,10 +76,12 @@ protected Task onResourceTick(AltoClef mod) {
 
         // No need to free inventory, output gets picked up.
 
-        setDebugState("Crafting in inventory... for " + toGet);
-        return mod.getModSettings().shouldUseCraftingBookToCraft()
-                ? new CraftGenericWithRecipeBooksTask(_target)
-                : new CraftGenericManuallyTask(_target);
+            setDebugState("Crafting in inventory... for " + toGet);
```
Comment (adrisj7) [inline src/main/java/adris/altoclef/tasks/CraftInInventoryTask.java:79]: Did this need to get indented?

File: a [hunk 14]
```diff
@@ -0,0 +1,487 @@
...
+Author: James Green <james@p.ckplace.org>
```
Comment (adrisj7) [inline a:2]: This file should be gitignored/removed from your branch to clean up this diff at least

File: src/main/java/adris/altoclef/tasks/ResourceTask.java [hunk 15]
```diff
@@ -176,6 +176,17 @@ protected Task onTick(AltoClef mod) {
             }
         }
 
+        return onResourceTick_clearingPhase(mod);
+    }
+
+    private Task onResourceTick_clearingPhase(AltoClef mod) {
```
Comment (adrisj7) [inline src/main/java/adris/altoclef/tasks/ResourceTask.java:182]: At this rate we should probably just move this logic up to `onTick`

File: src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+import adris.altoclef.AltoClef;
+import adris.altoclef.tasksystem.Task;
+import adris.altoclef.util.helpers.StorageHelper;
+import adris.altoclef.util.slots.PlayerSlot;
+import adris.altoclef.util.slots.Slot;
+import net.minecraft.item.ItemStack;
+import net.minecraft.resource.ResourceManager;
+
+public class EnsureFreeCraftingGridTask extends Task {
```
Comment (adrisj7) [inline src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java:11]: If it's just for the 2x2 grid maybe call it `Clear2x2CraftingGridTask` or something

File: src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+import adris.altoclef.AltoClef;
+
+/**
+ * Some tasks (mainly tasks that open containers) will break if they don't require
+ * that the 2x2 crafting grid is empty. Tasks which implement this interface
+ * require implementers to maintain an empty crafting grid.
+ *
+ * This interface let's a task declare that it MUST have certain slots clear before it can execute.
+ */
+public interface ITaskLimitsSlots {
```
Comment (adrisj7) [inline src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java:12]: I don't see why we need an interface, this can probably be deleted.

**Review by adrisj7 — CHANGES_REQUESTED** [review]
quick review stuff!

File: src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java (hunk 17 above)
Comment (JamesGreen31) [reply src/main/java/adris/altoclef/tasksystem/ITaskLimitsSlots.java]: Renamed and re-did better

File: src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java (hunk 16 above)
Comment (JamesGreen31) [reply src/main/java/adris/altoclef/tasks/slot/EnsureFreeCraftingGridTask.java]: done!

File: src/main/java/adris/altoclef/tasks/ResourceTask.java (hunk 15 above)
Comment (JamesGreen31) [reply src/main/java/adris/altoclef/tasks/ResourceTask.java]: done

File: src/main/java/adris/altoclef/tasks/CraftInInventoryTask.java (hunk 13 above)
Comment (JamesGreen31) [reply src/main/java/adris/altoclef/tasks/CraftInInventoryTask.java:79]: fixed

**Review by adrisj7 — APPROVED** [review]
Lookin good! >:)

**Pull request merged**